R/FilterStateChoices.R

```diff
@@ -1,16 +1,24 @@
+# ChoicesFilterState ------
+
 #' @name ChoicesFilterState
-#' @title `FilterState` object for factor or character variable
-#' @description Manages choosing elements from a set
 #' @docType class
-#' @keywords internal
 #'
+#' @title `FilterState` object for factor or character variable
+#'
+#' @description Manages choosing elements from a set.
 #'
 #' @examples
-#' filter_state <- teal.slice:::ChoicesFilterState$new(
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' ChoicesFilterState <- getFromNamespace("ChoicesFilterState", "teal.slice")
+#'
+#' filter_state <- ChoicesFilterState$new(
 #'   x = c(LETTERS, NA),
 #'   slice = teal_slice(varname = "x", dataname = "data")
 #' )
-#' shiny::isolate(filter_state$get_call())
```

Comment:
@pawelru I've seen you wanted to move `shiny` from Depends to Imports in `teal.slice` https://github.com/insightsengineering/teal.slice/issues/529
If we decide to do such a change, we should either put `shiny::` prefix in examples or add `library(shiny)` at the beginning of examples. Anyway I would opt for leaving `shiny::` prefix in examples due to this expected transition

Reply:
CC @kartikeyakirar

Reply:
Not required, we still have `@import shiny`.

Reply:
I though `@import shiny` is a `roxygen2` tag that will add a `import(shiny)` entry to NAMESPACE. This allows functions of teal.slice to be able to use functions from shiny package, but it does not allow global namespace to be able to use shiny functions. 

When you load `teal.slice`
```r
> library(teal.slice)
Loading required package: shiny
Warning message:
package ‘shiny’ was built under R version 4.3.1 
```
Shiny is loaded because it's in `Depends` field of the description, not because we have `@import shiny` roxygen2 tag.

We also have `import(R6)` but `R6` is only in `Imports` hence you can not directly call R6 functions

```r
> Queue <- R6Class("Queue",
+                  public = list(
+                    initialize = function(...) {
+                      for (item in list(...)) {
+                        self$add(item)
+                      }
+                    },
+                    add = function(x) {
+                      private$queue <- c(private$queue, list(x))
+                      invisible(self)
+                    },
+                    remove = function() {
+                      if (private$length() == 0) return(NULL)
+                      # Can use private$queue for explicit access
+                      head <- private$queue[[1]]
+                      private$queue <- private$queue[-1]
+                      head
+                    }
+                  ),
+                  private = list(
+                    queue = list(),
+                    length = function() base::length(private$queue)
+                  )
+ )
Error in R6Class("Queue", public = list(initialize = function(...) { : 
  could not find function "R6Class"
```

Even though examples may pass during the build without `library(shiny)` if it's only in Imports, I think when user runs examples in his/her section this will fail, THE MOMENT WE move shiny from Depends to Imports.

Reply:
> global namespace

What is that? :raised_eyebrow:

Reply:
A mind short cut of global `search()` path

Reply:
> Shiny is loaded because it's in `Depends` field of the description, not because we have `@import shiny` roxygen2 tag.

Wrong. It is _attached_ because it's in `Depends`, the default message is somewhat misleading. All it takes to load a namespace is to use `shiny::isolate` once.

We have `@import shiny` on top of `Depends: shiny` because that is demanded by checks: all packages in `Depends` must be imported from.

Reply:
I think we have `Depends: shiny` so that we do not need to put `library(shiny)` in every example that uses any `shiny` function (like `ShinyApp`). If we remove `shiny` from `Depends` you will need to bring back either `library(shiny)` or prefix every `shiny` function call with `shiny::`.

Running `shiny::isolate()` once does not allow to run other functions from `shiny` package without `library()` or prefixing.

```r
> shiny::isolate()
Error in contextFunc() : argument "expr" is missing, with no default
> shinyApp
Error: object 'shinyApp' not found
```

Reply:
> I think we have `Depends: shiny` so that we do not need to put `library(shiny)` in every example

You are mistaken. It is because `shiny` contains the core functionalities that `teal` builds upon.

Reply:
> Running `shiny::isolate()` once does not allow to run other functions from `shiny` package without `library()` or prefixing.

This is exactly the difference between loading and attaching.

Reply:
> P.S. IMO, all packages except teal could have shiny in imports.

In general I agree with this statement with exception of `teal` and also `teal.modules.xyz`. Feel free to discuss it here but I ddon't expect @kartikeyakirar to do this within this PR. That's why I created a separate issue (as opposed to a comment here) not to explode this PR.

Reply:
The issue https://github.com/insightsengineering/teal.slice/issues/529

Reply:
> You are mistaken. It is because shiny contains the core functionalities that teal builds upon.

agree, yeah but not for teal.slice

Reply:
>This is exactly the difference between loading and attaching.

You can’t attach a package without first loading it - `library()` loads then attaches the package.

Reply:
Let's remove `shiny::` and `library(shiny)` in examples, and let's bring them back if we decide to move `shiny` to `Imports`.

Reply:
> Let's remove shiny:: and library(shiny) in examples,

this is already done in all the document examples. I have removed two cases from the vignettes.[0f073f3](https://github.com/insightsengineering/teal.slice/pull/506/commits/0f073f33db5ebd3a819cdb211c9436fcaa7d986b)